research.rst

```diff
@@ -0,0 +1,315 @@
+Tequila
+=======
+
+What is Tequila?
+
+- it's a collection of Ansible roles
+- it's a wrapper script providing a simplified interface for the
+  ``ansible`` command-line program
+
+
+New Projects
+------------
+
+Creation of new projects would be done using django-project-template,
+just as we currently do for margarita.  Installing tequila would make
+available the ``tequila`` command, providing a simplified interface
+for ``ansible-playbook`` (and friends) that will be mostly familiar to
+current users of the django-project-template fabric scripts.  There
+would also be some mild changes to the project structure, i.e. the
+location of the project-specific variables and custom playbooks, since
+these would be kept distinct from paths where they might conflict with
+the equivalent Salt files.
+
+Much of this work has already been started, and is currently on the
+``use_tequila`` branch.
+
+
+Installation
+------------
+
+We need to have the ability to pin projects to particular versions of
+tequila.  It is also desirable to have an easy command to install or
+upgrade tequila, along the lines of ``pip install``.
+
+Under all of the proposals that follow, I recommend that Tequila be
+pip installable.  The work for this has already been done.
+
+
+Proposal 1: ``tequila roles`` command puts the roles in the right place
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Under this proposal, no changes to the structure of the Tequila
+repository need to happen.  Running the ``tequila roles`` command
+would symlink or copy the ``tequila/roles`` directory up to the
+current working directory (which under most scenarios would be the
+project directory), and this would generally not need to be run more
+than once per local setup.
+
+For development of the roles themselves, one could ``pip install -e
+<local_tequila_copy>``, and the in-progress versions of the roles
+would easily be made available for the ansible commands.
+
+The pros:
+
+- no changes to our Tequila repo structure
+- could still run plain versions of ``ansible-playbook``
+
+The cons:
+
+- our roles would not be easily usable by members of the Ansible
+  community
+
+
+Proposal 2: separate ``tequila-<rolename>`` repos
```

Comment:
I don't want to dismiss this without thinking about whether we could make it workable, but my initial reaction to this proposal is that it seems like a lot of complication and maintenance overhead.

I do think there'd be more benefits from keeping galaxy compatibility than just open-sourcing our work. It would impose a mindset of keeping things organized into well-differentiated roles with clearly defined boundaries, and make it less likely that we end up down the road with a big spaghetti of inter-dependent roles and tasks that nobody understands the relationships of.

How many roles would we be talking about here? If it's more than a handful or could grow to more than a handful, this is sounding like a lot to manage. And it doesn't appear that we can tell galaxy to automatically install some other galaxy roles when ours is installed, so any project using our roles would have to explicitly deal with all the top-level roles plus all the other roles they might depend on.  But if we can simplify things to just a few roles...

Would it help manage the complexity if we considered making some custom modules that our high-level roles could use, while still making the roles re-usable, since that's the whole point of not having this all in one repo?  Though again, there doesn't appear to be a good way to automate installing module dependencies either.

Switching back to reasons we might continue to use just one repository with multiple roles - in order to deploy more complicated environments, our roles are going to have interdependencies, and it might wind up that none of them are usable without the others anyway, even if we do keep them separated with well-defined boundaries.